File: 560-subarray-sum-equals-k/subarray-sum-equals-k.cpp
class Solution {
public:
    int subarraySum(vector<int>& nums, int k) {
        unordered_map<int, int> mp;
        mp[0] = 1;

        int currSum = 0;
        int count = 0;

        for(int num : nums){
            currSum += num;
            int req = currSum - k;
            if(mp.find(req) != mp.end()) count += mp[req];
            mp[currSum]++;
        }

        return count;
    }
};

File: 560-subarray-sum-equals-k/subarray-sum-equals-k.py
class Solution(object):
    def subarraySum(self, nums, k):
        prefix = {0 : 1}
        
        currSum = 0
        count = 0

        for num in nums:
            currSum += num
            needed = currSum - k

            if needed in prefix:
                count += prefix[needed]
            
            prefix[currSum] = prefix.get(currSum, 0) + 1
        
        return count
        